### PR TITLE
Add search view

### LINKED ISSUE #31: Add search view
When to 'search' click, icon color changes to green and a text field appears
Screenshot 2020-06-07 at 10 04 58
When you start typing, similar texts are searched throughout the whole site and need to show in which section it was found (Reports,  News, Money Gathering, etc)
Screenshot 2020-06-07 at 10 05 10
If a similar text is found, it is highlighting to green
Screenshot 2020-06-07 at 10 05 29
If there is no coincidence show such view
Screenshot 2020-06-07 at 10 05 36

### PR DESCRIPTION
**Pull request by R-Mosolov**
I had an intention to resolve this issue by using _minimum of code lines_ because such solution has easy supporting. 

**Summary:**
– the main functional requirement of Searcher component (searching an user information) has realised;
– but non-functional requirements have realised not full. For example, first letters of a search recommendation don't have green color yet. Also sections belonged to search suggestions don't appear into the separate block on right yet.
I am planning to fix these bugs when will appear more work practice with Vuetify.

Closes #31

The realization of this logic in native JavaScript you can see here: 
https://github.com/R-Mosolov/zakyat-frontend/tree/57fa232fab8e41b997e7477d0ab244a3bca13bd7/plugins/searcher

### CODE REVIEW

File: components/HeaderBar.vue [hunk 1]
```diff
@@ -50,7 +49,20 @@ export default Vue.extend({
   },
   methods: {
     rubles,
-    openSearcher: toggleSearcher,
+    toggleSearcher () {
```
Comment (enghamzasalem) [inline components/HeaderBar.vue:52]: Great Job Roman..
Better to use instance property From Vue js without local storage:
https://vuejs.org/v2/cookbook/adding-instance-properties.html
For example 
```suggestion
      data () {
    return {
      amount: 6234234,
      IsSearchBoxOpened: false,
    };
  },
  methods: {
    rubles,
    toggleSearcher () {
      const searchModalWindow = document.getElementById('searcher');
      if (searchModalWindow) {
        if (this.IsSearchBoxOpened) {
          searchModalWindow.style.display = 'none';
          this.IsSearchBoxOpened = false;
        } else {
          searchModalWindow.style.display = 'block';
          this.IsSearchBoxOpened = true;
        }
      }
    },
  },
```
image
Also, see https://vuejs.org/v2/guide/conditional.html maybe you will change your mind of using the display property

Comment (enghamzasalem) [reply components/HeaderBar.vue:52]: Read this interesting article about local storage https://dev.to/rdegges/please-stop-using-local-storage-1i04

File: components/HeaderBar.vue [hunk 2]
```diff
@@ -49,6 +49,20 @@ export default Vue.extend({
   },
   methods: {
     rubles,
+    toggleSearcher () {
+      const searchModalWindow = document.getElementById('searcher');
+      const isSearcherOpen = (localStorage.getItem('isSearcherOpen') === 'true');
+
+      if (searchModalWindow) {
```
Comment (aabounegm) [inline components/HeaderBar.vue:56]: Prefer to use guard clauses to decrease nesting
```suggestion
      if (!searchModalWindow) return;
```

File: components/HeaderBar.vue [hunk 3]
```diff
@@ -49,6 +49,20 @@ export default Vue.extend({
...
+        if (isSearcherOpen) {
+          searchModalWindow.style.display = 'none';
```
Comment (aabounegm) [inline components/HeaderBar.vue:58]: Instead of modifying the `style` directly, set a class that depends on the state, generally speaking. In this specific case when you are hiding/showing, use `v-show` or `v-if` instead.
Read more about it here: https://vuejs.org/v2/guide/conditional.html

File: components/common/Searcher.vue [hunk 4]
```diff
@@ -0,0 +1,59 @@
+<template>
+  <v-col id="searcher">
```
Comment (aabounegm) [inline components/common/Searcher.vue:2]: `v-col` is to be used wrapped with `v-row`. There is no direct parent `v-row` to this component. Also, don't set `id`s and use `ref`s instead

File: components/common/Searcher.vue [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    <v-autocomplete
+      id="searcher__input"
```
Comment (aabounegm) [inline components/common/Searcher.vue:4]: Another `id` that isn't used anywhere

File: components/common/Searcher.vue [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      :items="items"
+      background-color="white"
+      item-color="green"
+      placeholder="Введите фразу для поиска"
```
Comment (aabounegm) [inline components/common/Searcher.vue:8]: Don't hardcode any strings that can use i18n. Create a relevant entry in `ru.json` (and `en.json` if you can) and link it here.

File: components/common/Searcher.vue [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      no-data-text="Результатов нет. Попробуйте переформулировать."
```
Comment (aabounegm) [inline components/common/Searcher.vue:9]: Same thing applies here, but Vuetify's i18n strings are treated in a somewhat special way. You don't have to do it for now, but I'll leave this comment for me to remember.
https://vuetifyjs.com/en/customization/internationalization/#create-a-translation

File: components/common/Searcher.vue [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      append-icon=""
+      dense
+      clearable
+      rounded
+    ></v-autocomplete>
+  </v-col>
+</template>
+
+<script>
+export default {
```
Comment (aabounegm) [inline components/common/Searcher.vue:19]: We are using Typescript. To make sure everything is properly type-checked, use `lang="ts"` and `Vue.extend`
```suggestion
<script lang="ts">
import Vue from 'vue';

export default Vue.extend({
```

File: components/common/Searcher.vue [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        'Садака в размере 100 000 ₽ собрана за 10 дней в м... (Новости)',
+        'Садыков Амирхан (Сборы)',
+        'Садыков Амирхан (Отчеты)',
+      ],
+    };
+  },
+};
+</script>
+
+<style scoped>
```
Comment (aabounegm) [inline components/common/Searcher.vue:35]: Prefer to use Sass
```suggestion
<style lang="scss" scoped>
```

File: components/common/Searcher.vue [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+#list-24 {
+  padding-left: 10px;
+}
+
+.v-list-item {
+  font-weight: bold;
+  font-size: 16px;
+}
```
Comment (aabounegm) [inline components/common/Searcher.vue:58]: 1. Don't use the internal class used by the library. We don't control it and it can change any time.
2. Use Vuetify utility classes so that you don't need to even write CSS (https://vuetifyjs.com/en/styles/typography/)

File: components/common/Searcher.vue [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+  width: 600px;
+  z-index: 100;
+  background: white;
+  border-radius: 10px;
+  box-shadow: 0 4px 4px rgba(0, 0, 0, 0.25);
+}
+
+#list-24 {
+  padding-left: 10px;
+}
```
Comment (aabounegm) [inline components/common/Searcher.vue:53]: There is nothing with that `id`. Also, you can Vuetify's helper class `.pl-2`
https://vuetifyjs.com/en/styles/spacing/

**Comment by Yasinaaa** [comment]
Design bugs

Expectation
Screenshot 2020-06-11 at 14 02 58

Reality
Screenshot 2020-06-11 at 14 02 41

Fix: 
1) Remove harcode values from search view (Сабирзянова Римма Минзагитовна etc)
2) Where is close (x) button?

**Comment by Yasinaaa** [comment]
Design bugs. Part 2

Expectation
Screenshot 2020-06-11 at 14 04 08

Reality
Screenshot 2020-06-11 at 14 03 50


Fix:
1) Close button size. It's smaller than in Figma
2) Close button color must be green
3) The highlighted text must have a green color. It doesn't have a grey background
4) The found item type must not be in parentheses, it has to be in the right corner and with grey color
5) Wrong text style. Use Roboto font

**Comment by Yasinaaa** [comment]
Screenshot 2020-06-11 at 14 05 09

Screenshot 2020-06-11 at 14 04 36

Fix:
1) There are shouldn't be grey highlighted text. Just green text, check how is it created in Figma
2) Then you click to the found text, it shouldn't be paste to search view. It has to open the result page. Because we don't have such a page right now, just remove text from the search view and open any random link.

**Comment by R-Mosolov** [comment]
I have fixed 13 of 17 mistakes. I still didn't found ways to solve other mistakes. See details bellow.
First Code Review Result

File: components/HeaderBar.vue [hunk 12]
```diff
@@ -1,42 +1,54 @@
 <template>
-  <v-system-bar dark height="50" style="padding: 0 5%;">
+  <v-system-bar class="header-bar" height="50" style="padding: 0 5%;">
```
Comment (aabounegm) [inline components/HeaderBar.vue:2]: The `dark` prop already applies the style `background-color: black;`. Why did you change it?

File: components/HeaderBar.vue [hunk 13]
```diff
@@ -45,10 +57,48 @@ export default Vue.extend({
...
+<style lang="scss" scoped>
+  .header-bar {
+    background-color: black;
+  }
+
+  .header-bar button,
+  .header-bar .icon,
+  .descriptor {
+    color: white;
+  }
```
Comment (aabounegm) [inline components/HeaderBar.vue:79]: Didn't need to do this. `<v-icon>` supports the prop `color`, so you could apply `color="white"` directly on the icon

File: components/HeaderBar.vue [hunk 14]
```diff
@@ -45,10 +57,48 @@ export default Vue.extend({
...
+    background: white;
+    border-radius: 10px;
+    box-shadow: 0 4px 4px rgba(0, 0, 0, 0.25);
+    z-index: 2;
+  }
+
+  #list-24 {
+    padding-left: 10px;
+    font-family: "Roboto", Arial, sans-serif;
+  }
```
Comment (aabounegm) [inline components/HeaderBar.vue:103]: Still cannot find any element with that `id`

File: components/HeaderBar.vue [hunk 15]
```diff
@@ -1,42 +1,54 @@
...
       </v-icon>
       {{ $t('login') }}
     </v-btn>
     <v-flex lg1 md2 xs4>
-      <LanguageSelector />
+      <LanguageSelector class="icon" />
     </v-flex>
+    <v-col v-show="isSearcherOpen" class="searcher">
+      <v-autocomplete
+        class="searcher__input"
```
Comment (aabounegm) [inline components/HeaderBar.vue:35]: This class is not used in any CSS selector. It has no effect

**Comment by R-Mosolov** [comment]
> If you prefer, I can fix the remaining minor errors.
Yes, I agree. Could you sent me the notification after fixing remaining minor errors to learn methods that you use. Especially, methods of using slots in Vue are very interesting. This will useful for future development practice.

Am I understand correctly that I passed the test task? Can I take "Reports" page as #36 issue (see https://github.com/Zakyat/zakyat-frontend/issues/44) for realization following backlog element. It is good experience for learning how to work creating layouts, routing pages and filtering data in Vue.

**Comment by aabounegm** [comment]
You can check the commits to see the individual changes (and the commit message). I left the `"search.suggestions"` as it is for now because there is still a lot of logic that will go into implementing the actual search (frankly, I'm not even sure what it searches for and where to get that information from).
First 2 of my commits are cherry-picked from my commits on @stwer1998's branch (to avoid conflicts when merging both to master).
To answer your questions:
1. @Yasinaaa 
2. Also @Yasinaaa determines the priorities, but I don't have a problem with that
3. I'm fine with that

File: components/common/Searcher.vue (hunk 7 above)
Comment (R-Mosolov) [reply components/common/Searcher.vue]: Done.

File: components/HeaderBar.vue (hunk 13 above)
Comment (R-Mosolov) [reply components/HeaderBar.vue]: Done. Thank you for the suggestion. I will remember about this in the future.

File: components/HeaderBar.vue (hunk 14 above)
Comment (R-Mosolov) [reply components/HeaderBar.vue]: It isn't actual because we don't use `id` now. We use in the code Vue attributes instead. Thank you for the suggestion.

File: components/HeaderBar.vue (hunk 15 above)
Comment (R-Mosolov) [reply components/HeaderBar.vue]: The class has deleted. Thank you for attentiveness.
